aeneas/src/x86/X86MacroAssembler.v3

```diff
@@ -51,22 +73,51 @@ class X86MacroAssembler extends X86Assembler {
 		push(s);            // push value from source memory onto the stack
 		pop(X86Addr.!(d));  // pop value off stack into destination memory
 	}
+	// a macro to move between any two SSE register/memory operands
+	def movsd_sm_sm(d: SSERm, s: SSERm, scratch: SSEReg) {
+		if (s == d) return;
+		if (SSEReg.?(d)) return movsd_s_sm(SSEReg.!(d), s);
+		if (SSEReg.?(s)) return movsd_sm_s(d, SSEReg.!(s));
+		if (scratch != null) {
+			movsd_s_sm(scratch, s);
+			movsd_sm_s(d, scratch);
+		}
+		// push(s);            // push value from source memory onto the stack
+		// pop(SSEAddr.!(d));  // pop value off stack into destination memory
+	}
 	// a macro to move a value into a location
 	def movd_l_val(frame: MachFrame, loc: int, val: Val) {
-		var d = loc_rm(frame, loc);
-		if (val == null) {
-			movd_rm_i(d, 0);
-		} else if (Addr.?(val)) {
-			var pos = buffer.pos;
-			movd_rm_i(d, X86Addrs.ABS_CONST);
-			recordPatch(pos, Addr.!(val));
+		if (loc >= X86MachRegs.XMM0 && loc <= X86MachRegs.XMM7
+				|| Float32Val.?(val)) {
+			movd_l_fval(frame, loc, val);
 		} else {
-			movd_rm_i(d, V3.unboxIntegral(val));
+			var d = loc_rm(frame, loc);
+			match (val) {
+				x: Addr => {
+						 var pos = buffer.pos;
+						 movd_rm_i(d, X86Addrs.ABS_CONST);
+						 recordPatch(pos, x);
+				}
+				_ => movd_rm_i(d, V3.unboxIntegral(val));
+			}
 		}
 	}
+	// a macro to move a value into an SSE location
+	def movd_l_fval(frame: MachFrame, loc: int, val: Val) {
+		var f = 0;
+		var d = loc_sm(frame, loc);
+		if (val != null) {
+			match (val) {
+				x: Float32Val => f = int.view(x.bits);
+				x: Float64Val => f = int.view(x.bits);
```

Comment:
`int.view` will slice off only the low 32-bits. You'll have to handle doubles differently here.

Reply:
I have currently not implemented support for doubles. We can remove it for the time being, if you prefer.

Reply:
I've added a TODO as a reminder that it needs to be fixed.